### PR TITLE
Refactoring for kip103

### PR DESCRIPTION
**Pull request by yoomee1313**
## Proposed changes

- This PR is a refactoring for kip103.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [ ] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [ ] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [ ] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [ ] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

Referenced https://github.com/klaytn/klaytn/compare/dev...blukat29:klaytn:module-bc-system.

### CODE REVIEW

File: blockchain/system/rebalance.go [hunk 1]
```diff
@@ -163,7 +176,7 @@ func RebalanceTreasury(state *state.StateDB, chain consensus.ChainReader, header
 	totalRetiredAmount := result.totalRetriedBalance()
```
Comment (hyeonLewis) [inline blockchain/system/rebalance.go]:
```suggestion
	totalRetiredAmount := result.totalRetiredBalance()
```

File: blockchain/system/constant.go [hunk 2]
```diff
@@ -46,22 +46,42 @@ var (
...
+
 	// The following addresses are only used for testing.
 	Kip113ProxyAddrMock = common.HexToAddress("0x0000000000000000000000000000000000000402")
 	Kip113LogicAddrMock = common.HexToAddress("0x0000000000000000000000000000000000000403")
 
+	// TODO-Klaytn-Issue1587 Decide whether to cleanup below three addresses after discussing issue1587
+	RewardContractAddress = "0x0000000000000000000000000000000000000441"
+	KCFContractAddress    = "0x0000000000000000000000000000000000000442"
+	KFFContractAddress    = "0x0000000000000000000000000000000000000443"
+
```
Comment (blukat29) [inline blockchain/system/constant.go]: No need to bring them in. Let's just declare them in tests/klay_test_account_map_test.go.

File: consensus/istanbul/backend/engine.go [hunk 3]
```diff
@@ -531,8 +531,8 @@ func (sb *backend) Finalize(chain consensus.ChainReader, header *types.Header, s
 	if chain.Config().IsKIP103ForkBlock(header.Number) {
 		// RebalanceTreasury can modify the global state (state),
 		// so the existing state db should be used to apply the rebalancing result.
-		c := &Kip103ContractCaller{state, chain, header}
-		result, err := RebalanceTreasury(state, chain, header, c)
+		c := system.NewKip103ContractCaller(state, chain, header)
+		result, err := system.RebalanceTreasury(state, chain, header, c)
```
Comment (blukat29) [inline consensus/istanbul/backend/engine.go]: Please move the whole if clause inside blockchain/system. I want to see Finalize() eventually becoming a flat list of state transitions.

```
if chain.Config().IsKIP103ForkBlock(header.Number) {
  if err := system.ApplyKip103Transition(...); err != nil {
    return nil, err
  }
}
```

File: blockchain/system/rebalance_test.go [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+			ErrRebalanceBadStatus,
+			[]*big.Int{retirees[0].balance, retirees[1].balance, retirees[2].balance},
+			[]*big.Int{newbies[0].balance, newbies[1].balance},
+			9,
+			big.NewInt(0),
+			false,
+		},
+		{
+			// failed case - rebalancing aborted doe to bigger allocation than retirees
```
Comment (hyunsooda) [inline blockchain/system/rebalance_test.go]:
```suggestion
			// failed case - rebalancing aborted due to bigger allocation than retirees
```

File: blockchain/system/rebalance_test.go [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+		backend.Commit()
+		assert.Equal(t, uint64(1), backend.BlockChain().CurrentBlock().NumberU64())
+
+		// Get state and check post-transition state
+		state, err := chain.State()
+		assert.Nil(t, err)
+
+		header := chain.CurrentHeader()
+		c := &Kip103ContractCaller{state, chain, header}
```
Comment (hyunsooda) [inline blockchain/system/rebalance_test.go]:
```suggestion
		c := NewKip103ContractCaller(state, chain, header)
```

**Review by hyunsooda — DISMISSED** [review]

File: blockchain/system/rebalance.go (hunk 1 above)
Comment (yoomee1313) [reply blockchain/system/rebalance.go]: Thank you! reflected at https://github.com/klaytn/klaytn/pull/2121/commits/afa23a20492f5ec220a2cf3066bff98d0f5540c5

File: blockchain/system/rebalance_test.go (hunk 4 above)
Comment (yoomee1313) [reply blockchain/system/rebalance_test.go]: Thanks, reflected at https://github.com/klaytn/klaytn/pull/2121/commits/afa23a20492f5ec220a2cf3066bff98d0f5540c5

File: blockchain/system/rebalance_test.go (hunk 5 above)
Comment (yoomee1313) [reply blockchain/system/rebalance_test.go]: Thanks, reflected at https://github.com/klaytn/klaytn/pull/2121/commits/afa23a20492f5ec220a2cf3066bff98d0f5540c5

File: blockchain/system/constant.go (hunk 2 above)
Comment (yoomee1313) [reply blockchain/system/constant.go]: Sure, moved to test file. https://github.com/klaytn/klaytn/pull/2121/commits/657cf4218607bdc5d2bcf0cd6e0dc3dcaf3bfa5d

File: consensus/istanbul/backend/engine.go (hunk 3 above)
Comment (yoomee1313) [reply consensus/istanbul/backend/engine.go]: I moved the other code inside `TreasuryRebalance` code.
https://github.com/klaytn/klaytn/pull/2121/commits/de3f2205b7097d3cc5fc23b3c2e6dfe69dd916d3

File: blockchain/system/rebalance_test.go [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+		rebalanceBlockNumber *big.Int
+		status               uint8
+		newbieAmounts        []*big.Int
+
+		expectedErr           error
+		expectRetireesAmounts []*big.Int
+		expectNewbiesAmounts  []*big.Int
+		expectNonce           uint64
+		expectBurnt           *big.Int
+		expectSuccess         bool
```
Comment (ian0371) [inline blockchain/system/rebalance_test.go:73]: maybe add `expectedMemo`?

**Review by hyunsooda — DISMISSED** [review]

**Review by hyunsooda — DISMISSED** [review]

**Review by ian0371 — DISMISSED** [review]

**Comment by blukat29** [comment]
It has to be rebased after #2127.

**Comment by blukat29** [comment]
@yoomee1313 #2127 merged. Rebase please and sorry for the large changeset.

**Comment by yoomee1313** [comment]
This PR has been rebased. Please take a look. @blukat29 @ian0371 @hyeonLewis @hyunsooda

**Review by hyeonLewis — DISMISSED** [review]

**Comment by blukat29** [comment]
1. Perhaps you didn't modify `contracts/contracts/testing/system_contracts/all.sol`?
2. Do you intend to keep kip103 and upcoming kip160 contracts into `contracts/contracts/system_contracts/rebalance` altogether? If so, I am okay with the file relocation in this PR.

**Comment by yoomee1313** [comment]
> Perhaps you didn't modify contracts/contracts/testing/system_contracts/all.sol?

I missed it. Added the import statement of TreasuryRebalanceMock.sol at all.sol.

> Do you intend to keep kip103 and upcoming kip160 contracts into contracts/contracts/system_contracts/rebalance altogether? If so, I am okay with the file relocation in this PR

Correct. I'm planning to put kip103,kip160 contracts altogether in rebalance folder.

**Review by blukat29 — APPROVED** [review]
Thanks!

**Review by hyeonLewis — APPROVED** [review]

**Pull request merged**